classwork/classwork2.2/main.cpp

```diff
@@ -0,0 +1,150 @@
+        }
+        pixels.emplace_back(std::stoi(colorValues[0]), std::stoi(colorValues[1]), std::stoi(colorValues[2]));
+    }
+    return pixels;
+}
+
+// Function to process the image based on the favorite color
+void processImage(const std::string &inputFileName, const std::tuple<int, int, int> &favoriteColor, const std::string &outputFileName, const std::tuple<int, int, int> &unfavoriteColor = std::make_tuple(-1, -1, -1)) {
+    std::ifstream inputFile(inputFileName);
+    if (!inputFile.is_open()) {
+        throw std::runtime_error("Failed to open input file");
+    }
+
+    std::vector<std::vector<std::tuple<int, int, int>>> image(16, std::vector<std::tuple<int, int, int>>(16));
+
+    std::string line;
+    int row = 0;
+    while (std::getline(inputFile, line)) {
+        if (row >= 16) {
+            throw std::invalid_argument("Too many lines in input file");
+        }
+        image[row] = parseLine(line);
+        if (image[row].size() != 16) {
+            throw std::invalid_argument("Invalid number of pixels in a line");
+        }
+        ++row;
+    }
+    inputFile.close();
+
+    if (row != 16) {
+        throw std::invalid_argument("Not enough lines in input file");
+    }
+
+    for (int i = 0; i < 16; ++i) {
+        for (int j = 0; j < 16; ++j) {
+            if (image[i][j] == favoriteColor) {
+                if (i > 0 && j > 0) {
+                    image[i-1][j-1] = favoriteColor;
+                }
+            }
+            if (image[i][j] == unfavoriteColor) {
+                image[i][j] = favoriteColor;
+            }
+        }
+    }
+
+    std::ofstream outputFile(outputFileName);
+    if (!outputFile.is_open()) {
+        throw std::runtime_error("Failed to open output file");
+    }
+
+    for (const auto &row : image) {
+        for (size_t j = 0; j < row.size(); ++j) {
+            outputFile << std::get<0>(row[j]) << "," << std::get<1>(row[j]) << "," << std::get<2>(row[j]);
+            if (j < row.size() - 1) {
+                outputFile << " ";
+            }
+        }
+        outputFile << "\n";
+    }
+    outputFile.close();
+}
+
+int main(int argc, char *argv[]) {
+    if (argc != 6 && argc != 9) {
+        std::println("Usage: {} <input file> <R> <G> <B> <output file> [unfavorite R] [unfavorite G] [unfavorite B]\n", argv[0]);
+        return 1;
+    }
+
+    std::string inputFileName = argv[1];
+    int r = std::atoi(argv[2]);
+    int g = std::atoi(argv[3]);
+    int b = std::atoi(argv[4]);
+    std::string outputFileName = argv[5];
+
+    if (!isValidColorValue(r) || !isValidColorValue(g) || !isValidColorValue(b)) {
+        std::println("Error: Invalid favorite color value. Each component must be in the range [0, 255].\n");
+        return 1;
+    }
+
+    std::tuple<int, int, int> favoriteColor = std::make_tuple(r, g, b);
+
+    std::tuple<int, int, int> unfavoriteColor = std::make_tuple(-1, -1, -1);
+    if (argc == 9) {
+        int unfavoriteR = std::atoi(argv[6]);
```

Comment:
it looks like a copy paste of the block at line 117. I hope there should be a way to write a function that will return a Color object or std::optional<Color> based on input container or std::span on top of argv.